### ISSUE TITLE
lightgbm.cv not suport 'ndcg'?

### ISSUE THREAD

**Issue by deepfuzzy** [Microsoft/LightGBM#1502]
When I try the code in test_engine.py:

        # lambdarank
        X_train, y_train = load_svmlight_file(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../../examples/lambdarank/rank.train'))
        q_train = np.loadtxt(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../../examples/lambdarank/rank.train.query'))
        params_lambdarank = {'objective': 'lambdarank', 'verbose': -1}
        lgb_train = lgb.Dataset(X_train, y_train, group=q_train)
        lgb.cv(params_lambdarank, lgb_train, num_boost_round=10, nfold=3, stratified=False, metrics='l2', verbose_eval=False)
    
It's fine.

but when I change  metrics='l2' to 'ndcg', returns the message :
      'ndcg@1-mean': [nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]'
it's means cv not suport 'ndcg'? How to cv with 'ndcg'? 
thanks!

**Comment by guolinke** [comment]
@StrikerRUS I think we support this: https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L266-L273 
Any ideas?

**Comment by StrikerRUS** [comment]
@guolinke I confirm that this problem exists. Will try to find out why

**Comment by StrikerRUS** [comment]
the same happens with `map` metric

@guolinke Does `subset()` of `Dataset` class takes into account `group`?

also ping @wxchan

**Comment by guolinke** [comment]
@StrikerRUS 
subset used the indices generated by python, however, I forget it accept the new group or not.

**Comment by guolinke** [comment]
@StrikerRUS 
https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L290-L301

It seems the group isn't be assigned to subset

**Comment by StrikerRUS** [comment]
@guolinke 
>It seems the group isn't be assigned to subset

Yeah
Maybe this is the reason of `NaN`s?

```
import numpy as np
import lightgbm as lgb
from sklearn.datasets import load_svmlight_file

X_train, y_train = load_svmlight_file(r'C:\Users\nekit\Downloads\LightGBM/examples/lambdarank/rank.train')
q_train = np.loadtxt(r'C:\Users\nekit\Downloads\LightGBM/examples/lambdarank/rank.train.query')
params_lambdarank = {'objective': 'lambdarank', 'verbose': -1}
lgb_train = lgb.Dataset(X_train, y_train, group=q_train).construct()
lgb_train.get_group()

>>> array([  1.,  13.,   5.,   8.,  19.,  12.,  18.,   5.,  14.,  13.,   8.,
         9.,  16.,  11.,  21.,  14.,  21.,   9.,  14.,  11.,  20.,  18.,
        13.,  20.,  22.,  22.,  13.,  17.,  10.,  13.,  12.,  13.,  13.,
        23.,  18.,  13.,  20.,  12.,  22.,  14.,  13.,  23.,  13.,  14.,
        14.,   5.,  13.,  15.,  14.,  14.,  16.,  16.,  15.,  21.,  22.,
        10.,  22.,  18.,  25.,  16.,  12.,  12.,  15.,  15.,  25.,  13.,
         9.,  12.,   8.,  16.,  25.,  19.,  24.,  12.,  16.,  10.,  16.,
         9.,  17.,  15.,   7.,   9.,  15.,  14.,  16.,  17.,   8.,  17.,
        12.,  18.,  23.,  10.,  12.,  12.,   4.,  14.,  12.,  15.,  27.,
        16.,  20.,  13.,  19.,  13.,  17.,  17.,  16.,  12.,  15.,  14.,
        14.,  19.,  12.,  23.,  18.,  16.,   9.,  23.,  11.,  15.,   8.,
        10.,  10.,  16.,  11.,  15.,  22.,  16.,  17.,  23.,  16.,  22.,
        17.,  14.,  12.,  14.,  20.,  15.,  17.,  15.,  15.,  22.,   9.,
        21.,   9.,  17.,  16.,  15.,  13.,  13.,  15.,  14.,  18.,  21.,
        14.,  17.,  15.,  14.,  16.,  12.,  17.,  19.,  16.,  11.,  18.,
        11.,  13.,  14.,   9.,  16.,  15.,  16.,  25.,   9.,  13.,  22.,
        16.,  18.,  20.,  14.,  11.,   9.,  16.,  19.,  19.,  11.,  11.,
        13.,  14.,  14.,  13.,  16.,   6.,  21.,  16.,  12.,  16.,  11.,
        24.,  12.,  10.])

subset = lgb_train.subset([0,1,2,3,4,5,6,7,8,9]).construct()
subset.get_group()

>>> []
```

**Comment by guolinke** [comment]
The subset's api only support using indices now. 
You can use Dataset.set_group to set group to these subsets.

**Comment by StrikerRUS** [comment]
OK, can we then subset `group` this way?
```diff
D:\Users\nekit\Downloads\LightGBM>git diff
diff --git a/python-package/lightgbm/basic.py b/python-package/lightgbm/basic.py
index 33dda5e..c44953a 100644
--- a/python-package/lightgbm/basic.py
+++ b/python-package/lightgbm/basic.py
@@ -888,6 +888,9 @@ class Dataset(object):
                     # construct subset
                     used_indices = list_to_1d_numpy(self.used_indices, np.int32, name='used_indices')
                     assert used_indices.flags.c_contiguous
+                    if self.group is not None:
+                        group_info = np.array(self.group).astype(int)
+                        _, self.group = np.unique(np.repeat(range_(len(group_info)), repeats=group_info)[self.used_indices], return_counts=True)
                     self.handle = ctypes.c_void_p()
                     params_str = param_dict_to_str(self.params)
                     _safe_call(_LIB.LGBM_DatasetGetSubset(
@@ -959,7 +962,7 @@ class Dataset(object):
         """
         if params is None:
             params = self.params
-        ret = Dataset(None, reference=self, feature_name=self.feature_name,
+        ret = Dataset(None, reference=self, feature_name=self.feature_name, group=self.group,
                       categorical_feature=self.categorical_feature, params=params)
         ret._predictor = self._predictor
         ret.pandas_categorical = self.pandas_categorical
```

**Comment by StrikerRUS** [comment]
After this patch, two last lines of this snippet https://github.com/Microsoft/LightGBM/issues/1502#issuecomment-406567016

```
subset = lgb_train.subset([0,1,2,3,4,5, 155,156, 200]).construct()
subset.get_group()
```

yields
```
array([1, 5, 2, 1], dtype=int64)
```

**Comment by guolinke** [comment]
@StrikerRUS Is that correct? The data belongs the same group should be in the same subset. 
Maybe we should check the indices when subset with group.

**Comment by StrikerRUS** [comment]
I haven't touched subsetting logic. It still subsets by data indices. I've just reorganized `group` according to new number of samples, but remembering previous `group`.

Let say we have dataset with 19 samples and `group` is `[1, 14, 5]`.
After subsetting 
```
dataset.subset([0,    1,2,5,7,     15,17])
```
new group will be `[1, 4, 2]`.

It is correct?

**Comment by guolinke** [comment]
@StrikerRUS yeah, it is.

**Comment by Laurae2** [comment]
@guolinke in R it passes the group correctly I think. It was one of the issues I had to fix when reworking lgb.cv (not sure which PR / issue).

**Comment by guolinke** [comment]
@StrikerRUS 
I think there is something wrong when calling ndcg/map. otherwise it will throw a error when without group: https://github.com/Microsoft/LightGBM/blob/master/src/metric/rank_metric.hpp#L47

**Comment by StrikerRUS** [comment]
@guolinke Strange...

As we identified, `subset()` method doesn't preserve `group` a present. However, `cv` function uses subsetting:
https://github.com/Microsoft/LightGBM/blob/225eb5eddac3fd26f3717cfbec88cfcf694dcbc8/python-package/lightgbm/engine.py#L291-L293

**Comment by guolinke** [comment]
maybe there is something wrong with dataset.construct. I will find a time to check this.

**Comment by guolinke** [comment]
@StrikerRUS do you have the branch of your fixes ? I want test on it, thanks

**Comment by StrikerRUS** [comment]
@guolinke Unfortunately, all changes were done locally and only this snippet are left https://github.com/Microsoft/LightGBM/issues/1502#issuecomment-408610094

**Comment by guolinke** [comment]
@StrikerRUS can you help to test `ndcg-cv` branch ?

**Comment by StrikerRUS** [comment]
@guolinke Sure! Will do it today.

**Comment by guolinke** [comment]
fixed in #1646
